### ISSUE TITLE
Support different handlers for different reverse-proxied ports

### ISSUE THREAD

**Issue by edk0** [paramiko/paramiko#614]
[The documentation](http://docs.paramiko.org/en/1.16/api/transport.html#paramiko.transport.Transport.request_port_forward) for `Transport.request_port_forward` (as well as the design of its interface, IMO) suggests that different forwarded ports should be able to have different handlers associated with them; in fact there's just one instance attribute, `_tcp_handler`, which is set every time a new forward is requested.

I'm not sure exactly what to do instead—probably we want a mapping of active port forward requests to handlers, and dispatch on that for new forwarded channels?

Probably relevant to #544. Also possibly relevant to X11, but I have no idea how X11 forwarding works anyway.

**Comment by cnelson** [comment]
Yes, I'd appreciate feedback on this from what would be acceptable in a PR.  I need the UDP support mentioned in #544 and have my own working patch, but it seems like it might be appropriate to make the fix more generic as suggested here.
